### PR TITLE
Adds our own Cloud Functions sanitization library + Sanitizes all CFs

### PR DESCRIPTION
**Pull request by holoisme**
Hello,

As we've seen in our Computer Security class, lack of sanitization in arguments is the root of a lot of issues and mainly security breaches, this PR solves this issue by providing a fully featured sanitization library for our Cloud Functions backend.

It's relatively easy for a hacker to extract the `google_services.json` file from an `.apk`, which would then let him call our HTTP end points (Cloud Functions) with any arguments. It is a good thing to keep in mind that **Typescript doesn't actually exist in runtime**, so there is no type checking when the functions are deployed, Typescript only exists in our code editors at compile time.

## How to use it
The main building block that we will be using all the time is `Predicate`. A [precidate](https://en.wikipedia.org/wiki/Predicate_(mathematical_logic)) is simply a generic function that takes in a value and returns a `boolean`.
We will use those predicates to enforce all the rules that the arguments **must** follow for the function to actually do its job and prevent hackers from abusing it.

To integrate the sanitization seamlessly and in the most intuitive way, we will now use `onSanitizedCall` rather than `onAuthenticatedCall`, the structure is very similar. Here is how to use it :
```typescript
export const myFunction = onSanitizedCall(Path.To.myFunction, {
    /* sanitization here, we define the predicates on all the fields of our arguments */
}, (userId, args) => {
    /* business logic here, as usual */
})
```

Let's imagine I want to enforce the following function arguments :
```typescript
interface FunctionArgs {
    name: string,
    age: number,
    gender: "m" | "f",
    nickname?: string,
    socialSecurityNumbers: number[]
}
```
Then we will use the following predicates :
```typescript
{
    name: Predicate.isString,
    age: Predicate.isPositive // This of course also checks if it is a number, it is contingent
    gender: Predicate.isIn(["m", "f"] as const),
    nickname: Predicate.isOptionalString,
    socialSecurityNumbers: [Predicate.isNonEmptyArray, Predicate.forEach(Predicate.isPositive)]
}
```
We have a perfect example of `Predicate` composition in this last field `socialSecurityNumbers` :
- **An array of predicates is itself a predicate**, and it returns true only if all of its predicates are satisfied.
- Since `socialSecurityNumbers` is an array, we can use the predicate `forEach` which takes in another predicate, and all of the elements of that array must satisfy the given predicate.

This library also gives full type checking for Intellisense (you can't put a predicate `isNumber` on a `string` field for example) which prevents us from making silly mistakes.

**The case disjunction** is a very common pattern in our model, let's look at how to handle it :
```typescript
export interface MCQ {
    type: "MCQ",
    question: string,
    imageURL?: string,
    propositions: string[],
    answersIndices: number[],
    numberOfAnswers: number
}

interface TF {
    type: "TF",
    question: string,
    imageURL?: string,
    answer: boolean
}

type Exercise = MCQ | TF;

interface FunctionArgs {
    exercises: Exercise[]
}
```
We can see here that `Exercise` has a discriminant key `type` which can either take `MCQ` or `TF`, the sanitization of `FunctionArgs` would then be the following :
```typescript
exercises: Predicate.forEach(
    Predicate.dispatch("type", {
        MCQ: Predicate.fields({
            answersIndices: Predicate.forEach(Predicate.isNumber),
            imageURL: Predicate.isOptionalString,
            numberOfAnswers: Predicate.isStrictlyPositive,
            propositions: Predicate.forEach(Predicate.isString),
            question: Predicate.isString
        }),
        TF: Predicate.fields({
            answer: Predicate.isBoolean,
            question: Predicate.isString,
            imageURL: Predicate.isOptionalString
        })
    })
)
```

These were the relatively simple examples, there are much more advanced examples in the sanitized version of the cloud functions this PR provides.

If you want to create custom predicates, such as `isValidExercise` for example, check `utils/custom-sanitizer` to see examples of how to do so cleanly.

Good luck and stay safe with your functions!

### CODE REVIEW

File: edweiss-firebase/functions/src/utils/firebase.ts [hunk 1]
```diff
@@ -16,7 +17,37 @@ export function onAuthentifiedCall<Args, Result, Error>(
...
+				return e;
+
+			return INTERNAL_ERROR;
+		}
+	});
+
+	return { signature, handler };
+}
+
+export function onSanitizedCall<Args extends object, Result, Error>(
```
Comment (FerdinandGomerChiu) [inline edweiss-firebase/functions/src/utils/firebase.ts:33]: What does this function achieve? I saw that you replaced an onAuthentifiedCall with it earlier (createQuiz function).

Comment (holoisme) [reply edweiss-firebase/functions/src/utils/firebase.ts:33]: Hello,
This achieves practically the same thing as `onAuthentifiedCall`, it simply checks if the argument predicates are satisfied before the callback.
But you're right, I forgot to document it.

Comment (FerdinandGomerChiu) [reply edweiss-firebase/functions/src/utils/firebase.ts:33]: I see. Good job as always, this is an absolutely bonkers amount of work! :+1:

**Comment by FerdinandGomerChiu** [comment]
Consider adding a guide to these assertions in the wiki. This is a lot of new content! (which is a great problem to have, don't you think?)

**Comment by holoisme** [comment]
> Consider adding a guide to these assertions in the wiki. This is a lot of new content! (which is a great problem to have, don't you think?)

Oh you're right! Forgot this existed, thanks for the idea!

**Review by FlorianDinant — APPROVED** [review]
Hello, I’ve taken a deep look at the changes, and I can see the effort you've done for making the code more readable and more secure. You have really done a good job!

To be honest, there's not much to say about potential change requests. Your PR is very good! However, I'd like to take the liberty of summarizing what I've observed in the modified files, and possibly check certain points, even if in reality everything seems correct.

## Summary of Changes:
### I've noticed the following updates:
- You renamed `onAuthentifiedCall` with `onSanitizedCall`.
- You added `Predicate` import for validation.

### In `edweiss-firebase/functions`
#### In the _`package-lock.json`_ file:
- Upgraded typescript from version `4.9.0` to `5.6.3`.
- Updated the resolved URL and integrity hash for typescript.
- Changed the required Node.js engine version for typescript from `>=4.2.0` to `>=14.17`.

Have you ensured the compatibility with the new TypeScript and Node.js versions across the project to avoid any unexpected issues?

#### In the _`package.json`_ file:
- Upgraded typescript from version `4.9.0` to `5.6.3`.

Make sure also that it is compatible across the project if you haven't already done so.

#### In the _`tsconfig.json`_ file:
- Updated the `target` from `es2017` to `ES2020`. This change allows the use of newer JavaScript features and improvements introduced in ES2020.


### In `functions/action/`

#### In the _`createQuiz.ts`_ file:
- Added validation predicates for `courseId` and `quiz`.
- Removed manual checks for `empty quiz.exercises` and `quiz.name`.

Have you ensured that the new validation predicates adequately cover all necessary checks to prevent invalid data from being processed ?

#### In the _`createQuizAttempt.ts`_ file:
- Added validation predicates for `courseId`, `path`, `quizId`, and `quizAttempt`.
- Removed commented-out code that checked if `userId` is undefined.

#### In the _`updateQuiz.ts`_ file:
- Added validation predicates for `courseId` and `quiz`.
- Introduced assertions `assertNonEmptyString` and `assertThat` for `courseId` and `quiz`.
- Removed commented-out code checking user authorization.

### In `functions/auth/`
#### In the _`createAccount.ts`_ file:
- Added  `tag` imports for validation.
- Added validation predicate for `name` using `tag(Predicate.isNonEmptyString, 'invalid_name')`.

### In `functions/experiments/`

#### In the _`promptAI.ts`_ file:
- Added validation predicates for task and content.
- Removed the try-catch block and the call to fail.
- Added a comment indicating that this is an experimental function.

### In `functions/fcm/`

#### In the _`registerFCMToken.ts`_ file:
- Added validation predicate for `fcmToken`.
- Removed manual checks for `userId` and `fcmToken`.

#### In the _`sendFCMPage.ts`_ file:
- Added validation predicate for `page` using `Predicate.isPositive`.
- Removed manual checks for `userId` and `page`.

### In `functions/lectures/`

#### In the _`addAudioTranscript.ts`_ file:
- Added validation predicates for `courseId`, `lectureId`, `transcription`, and `pageNumber`.
- Added `assert` imports for validation.
- Removed manual checks for arguments and replaced with assertions.
- Updated the return statement to use `ok('successfully_added')` and commented that "successfully_added" is a useless information.

If not done yet, please ensure that the new validation predicates and assertions adequately cover all necessary checks to prevent invalid data from being processed.

### In `functions/memento/`
#### In the _`createCard.ts`_ file:
- Added validation predicates for `deckId` and `card` fields including `answer`, `question`, and `learning_status`.

#### In the _`createDeck.ts`_ file:
- Added validation predicates for `deck` fields including `name` and `cards` using `CustomPredicateMemento.isValidCard`.

If not done yet, please ensure that the new validation predicates adequately cover all necessary checks to prevent invalid data from being processed. The use of custom predicates should be verified to ensure they meet the required validation logic.

#### In the _`deleteCard.ts`_ file:
- Added `Predicate` import for validation.
- Added validation predicates for `cardIndex` and `deckId`.

#### In the _`deleteDeck.ts`_ file:
- Introduced a validation predicate to ensure `deckId `is a non-empty string.
- Removed the manual check for `deckId`, replaced with an automatic validation via `Predicate`.

#### In the _`updateCard.ts`_ file:
- Introduced validation predicates for `cardIndex`, `deckId`, and `newCard` using custom predicates for card validation.
- Modified the function to update the card within the deck using the sanitized arguments.

#### In the _`updateDeck.ts`_ file:
- Removed an unnecessary blank line for better readability.
- Included a placeholder for future implementation.

### In `functions/notifs/`
#### In the _`deleteNotif.ts`_ AND the _`pushNotif.ts`_ files:
- Added `assertNonEmptyString` import for input validation.
- Replaced manual check for `args.id` with `assertNonEmptyString` to ensure the` id` is a non-empty string.

#### In the _`markAsRead`_ AND the _`markAsUnread`_ files:
- Added `assertNonEmptyString` import for input validation.
- Replaced manual check for `args.id` with `assertNonEmptyString` to ensure the `id` is a non-empty string.

### In functions/todolist/
#### In the _`createTodo.ts`_ file:
- Introduced validation predicates for `name`, `status`, `description`, and `dueDate` using custom predicates for validation.
- Modified the function to create a to-do item using the sanitized arguments.

#### In the _`deleteTodo.ts`_ file:
- Introduced a validation predicate for `id` to ensure it is a non-empty string.
- Removed the try-catch block and manual check for `id`, replaced with an automatic validation via `Predicate`.
- Updated the return statement to use `OK` instead of `ok({})`.

#### In the _`updateTodo.ts`_ file:
- Added `assertNonEmptyString` imports for validation purposes.
- Introduced validation predicates for` id`, `description`, `dueDate`, `name`, and `status` using custom predicates for validation.
- Modified the function to update a to-do item using the sanitized arguments.

### In `src/utils/`
#### In the _`firebase.ts`_ file:
- **Imports:**
   - Added `FailedCallResult` import from `model/functions` for handling failed call results.
   - Added `Predicate`, `PredicateOnFields`, and `assertThat` imports from `sanitizer` for input validation.
   - Added `INTERNAL_ERROR` import from `status` for handling internal errors.

- **onAuthentifiedCall Function:**
   - Added a try-catch block to handle potential errors during the callback execution.
   - Returns `INTERNAL_ERROR` if an unexpected error occurs.

- **onSanitizedCall Function:**
   - New function introduced that extends `onAuthentifiedCall`.
   - Adds argument validation using predicates before executing the callback.
   - Uses `assertThat` to validate the request data against the provided predicates.
   - Handles potential errors during validation and callback execution, returning `INTERNAL_ERROR` if an unexpected error occurs.

#### In the _`sanitizer.ts`_ file:
- **Imports:**
   - Added `FailedCallResult` from `model/functions` to handle failed call results.
   - Added `INVALID_ARGUMENT` and `fail` from `status` for error handling.

- **Types:**
   - Defined `Predicate`, `CompositePredicate`, and `PredicateOnFields` for type-safe input validation.
   - Introduced a new type `AssertionError` for handling assertion errors.
- **Functions:**
   - **throwable**: Converts an error into a `FailedCallResult`.
   - **taggedThrowable**: Tags an error with a specific message for debugging.

- **Assertions:**
   - **assert**: Throws an error if a condition is false.
   - **assertThat**: Throws an error if a value does not satisfy a predicate.
   - **assertThatFields**: Validates multiple fields against their respective predicates.
   - **assertNumber**: Ensures a value is a number.
   - **assertString**: Ensures a value is a string.
   - **assertBoolean**: Ensures a value is a boolean.
   - **assertObject**: Ensures a value is an object.
   - **assertArray**: Ensures a value is an array.
   - **assertPositiveNumber**: Ensures a value is a positive number.
   - **assertStrictlyPositiveNumber**: Ensures a value is a strictly positive number.
   - **assertIsBetween**: Ensures a value is between a specified range.
   - **assertNonEmptyString**: Ensures a string is not empty.
   - **assertNonEmptyStrings**: Ensures multiple strings are not empty.
   - **assertIsIn**: Ensures a value is in a specified list.
   - **assertForEach**: Ensures all elements of an array satisfy a predicate.

- **Tagging:**
   - **tag**: Tags a predicate with an error message for debugging purposes.

- **Predicate Namespace:**
   - Provided various utility predicates for common validation tasks, such as checking if a value is defined, a number, a string, an array, positive, non-empty, etc.
   - **dispatch**: Handles case disjunction based on a key.
   - **fields**: Validates all fields of an object against their respective predicates.
   - **not**: Inverts the result of predicates.
   - **verify**: Checks if a value satisfies a composite predicate.
   - **toPredicate**: Converts an array of composite predicates into a simple predicate function.

**Recommendation:**
- Ensure that all predicates and assertions are correctly implemented and cover all necessary validation cases. The tagging functionality should be verified for its usefulness in debugging.

#### In the _`status.ts`_ file:
- **New Constants:**
   - Added `INVALID_ARGUMENT` constant to represent an invalid argument error.
   -Added `INVALID_COURSE_ID` constant to represent an invalid course ID error.

- **Existing Constants:**
   - The existing constants `FAILED`, `NOT_AUTHENTIFIED`, and `INTERNAL_ERROR` remain unchanged.

**Recommendation:**
- Ensure that the new error constants `INVALID_ARGUMENT` and `INVALID_COURSE_ID` are correctly integrated and used appropriately in the application to handle specific error cases.

### In `utils/custom-sanitizer/`:
#### In the _`memento.ts`_ file:
- **Imports:**
   - Imported `Memento` from `model/memento`.
   - Imported `Predicate` from `utils/sanitizer`.

- **CustomPredicateMemento Namespace:**
   - Defined a namespace `CustomPredicateMemento` to encapsulate custom predicates for `Memento`.

- **isValidCard Predicate:**
   - Added `isValidCard` predicate to validate `Memento.Card` objects.
   - Checks that `answer` and `question` fields are non-empty strings.
   - Ensures that `learning_status` is either "Got it", "Not yet", or undefined (optional).

**Recommendation:**
- Ensure that the `isValidCard` predicate covers all necessary validation rules for `Memento.Card` objects and verify that it is used appropriately in the application.

#### In the _`quiz.ts`_ file:
- **Imports:**
   - Imported `Predicate` from `utils/sanitizer`.
   - Imported `Quizzes` and `QuizzesAttempts` from `model/quizzes`.

- **CustomPredicateQuiz Namespace:**
   - Defined a namespace `CustomPredicateQuiz` to encapsulate custom predicates for `Quizzes` and `QuizzesAttempts`.

- **isValidQuiz Predicate:**
   - Added `isValidQuiz` predicate to validate `Quizzes.Quiz` objects.
   - Checks that` answers` is a non-empty array and each item satisfies the specified dispatch predicates.
   - Ensures `deadline` is an optional string.
   - Ensures `ended` and `showResultToStudents` are booleans.
   - Checks that `exercises` is a non-empty array and each item satisfies the specified dispatch predicates.
   - Ensures `name` is a non-empty string.

- **isValidQuizAttempt Predicate:**
   - Added `isValidQuizAttempt` predicate to validate `QuizzesAttempts.QuizAttempt` objects.
   - Ensures `attempts` is a positive number.
   - Checks that `answers` is a non-empty array and each item satisfies the specified dispatch predicates.

**Recommendation:**
- Ensure that the `isValidQuiz` and `isValidQuizAttempt` predicates cover all necessary validation rules for `Quizzes.Quiz` and `QuizzesAttempts.QuizAttempt `objects respectively, and verify that they are used appropriately in the application.

## Conlusion: 
The changes in this pull request effectively upgrade TypeScript, improve code sanitization, and refactor various functions to enhance security and maintainability. Excellent work on implementing these updates and maintaining code quality. Congratulations on a job well done!

**Pull request merged**